services/galley/src/Galley/Cassandra/MakeFeature.hs

```diff
@@ -0,0 +1,449 @@
+    Tuple (TupleP mrow)
+  ) =>
+  TeamId ->
+  m (DbFeature cfg)
+fetchFeature tid = do
+  case featureColumns @cfg of
+    Nil -> pure (rowToFeature Nil)
+    cols -> do
+      let select :: PrepQuery R (Identity TeamId) (TupleP mrow)
+          select =
+            fromString $
+              "select "
+                <> intercalate ", " (hcollapse cols)
+                <> " from team_features where team_id = ?"
+      row <- retry x1 $ query1 select (params LocalQuorum (Identity tid))
+      pure $ foldMap (rowToFeature . unfactorI . productTypeFrom) row
+
+storeFeature ::
+  forall cfg m row mrow.
+  ( MonadClient m,
+    row ~ FeatureRow cfg,
+    MakeFeature cfg,
+    IsProductType (TupleP (TeamId : mrow)) (TeamId : mrow),
+    AllZip (IsF Maybe) row mrow,
+    Tuple (TupleP (TeamId : mrow)),
+    KnownNat (Length row)
+  ) =>
+  TeamId ->
+  LockableFeature cfg ->
+  m ()
+storeFeature tid feat = do
+  if n == 0
+    then pure ()
+    else
+      retry x5 $
+        write
+          insert
+          ( params LocalQuorum (productTypeTo (I tid :* factorI (featureToRow feat)))
+          )
+  where
+    n :: Int
+    n = fromIntegral (demote @(Length row))
+
+    insert :: PrepQuery W (TupleP (TeamId ': mrow)) ()
+    insert =
+      fromString $
+        "insert into team_features (team_id, "
+          <> intercalate ", " (hcollapse (featureColumns @cfg))
+          <> ") values ("
+          <> intercalate "," (replicate (succ n) "?")
+          <> ")"
+
+class (FeatureRow cfg ~ row) => StoreFeatureLockStatus (row :: [Type]) cfg where
+  storeFeatureLockStatus' :: (MonadClient m) => TeamId -> Tagged cfg LockStatus -> m ()
+
+instance
+  {-# OVERLAPPING #-}
+  ( FeatureRow cfg ~ (LockStatus ': row),
+    MakeFeature cfg
+  ) =>
+  StoreFeatureLockStatus (LockStatus ': row) cfg
+  where
+  storeFeatureLockStatus' tid lock = do
+    let col = unK (hd (featureColumns @cfg))
+        insert :: PrepQuery W (TeamId, LockStatus) ()
+        insert =
+          fromString $
+            "insert into team_features (team_id, " <> col <> ") values (?, ?)"
+    retry x5 $ write insert (params LocalQuorum (tid, (untag lock)))
+
+instance (FeatureRow cfg ~ row) => StoreFeatureLockStatus row cfg where
+  storeFeatureLockStatus' _ _ = pure ()
+
+storeFeatureLockStatus ::
+  forall cfg m.
+  (MonadClient m, StoreFeatureLockStatus (FeatureRow cfg) cfg) =>
+  TeamId ->
+  Tagged cfg LockStatus ->
+  m ()
+storeFeatureLockStatus = storeFeatureLockStatus' @(FeatureRow cfg)
+
+-- | Convert @NP f [x1, ..., xn]@ to @NP I [f x1, ..., f xn]@.
+--
+-- This works because @I . f = f@.
+factorI :: forall f xs ys. (AllZip (IsF f) xs ys) => NP f xs -> NP I ys
+factorI Nil = Nil
+factorI (x :* xs) = I x :* factorI xs
+
+-- | Convert @NP I [f x1, ..., f xn]@ to @NP f [x1, ..., xn]@.
+--
+-- See 'factorI'.
+unfactorI :: forall f xs ys. (AllZip (IsF f) xs ys) => NP I ys -> NP f xs
+unfactorI Nil = Nil
+unfactorI (I x :* xs) = x :* unfactorI xs
```

Comment:
it might be possible to do this with `htrans`. I thought about it a bit more which made me wonder why this is not in the library but I think it is. This seems to be exactly like htrans in that given a constraint that relates the element types (this is `IsF f`) you can make a natural transformation on the functor. I thought that `hfromI` or `htoI` would be the ones we want but it's not quite that because they require a `Coercible` constraint on `I` and `f` which is not what we want.